relay-server/src/middlewares/body_timing.rs

```diff
@@ -0,0 +1,282 @@
+use crate::statsd::RelayTimers;
+use axum::body::{Body, HttpBody};
+use axum::http::Request;
+use hyper::body::Frame;
+use relay_statsd::metric;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::time::Instant;
+use tower::{Layer, Service};
+
+/// Middleware layer that wraps the request [`Body`] to measure body reading duration.
+///
+/// Each metric includes the following tags:
+/// - `route`: The request's URI path
+/// - `size`: Size category based on bytes read
+/// - `status`: Reading outcome, one of:
+///   - `"completed"`: Body was fully read
+///   - `"failed"`: Reading encountered an error
+///   - `"dropped"`: Body was dropped before completion
+///
+/// The size reflects actual bytes read, which may differ from the `Content-Length` header.
+/// For interrupted reads (failed/dropped), only successfully read bytes are counted.
+#[derive(Clone, Debug, Default)]
+pub struct BodyTimingLayer;
+
+impl<S> Layer<S> for BodyTimingLayer {
+    type Service = BodyTiming<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        BodyTiming::new(inner)
+    }
+}
+
+/// Tower service provided by [`BodyTimingLayer`].
+#[derive(Debug, Copy, Clone)]
+pub struct BodyTiming<S> {
+    inner: S,
+}
+
+impl<S> BodyTiming<S> {
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S> Service<Request<Body>> for BodyTiming<S>
+where
+    S: Service<Request<Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let (parts, body) = req.into_parts();
+        let timed_body = Body::new(TimedBody::new(body, parts.uri.to_string()));
+        let request = Request::from_parts(parts, timed_body);
+        self.inner.call(request)
+    }
+}
+
+#[derive(Debug)]
+struct TimedBody {
+    inner: Body,
+    reading_started_at: Option<Instant>,
+    size: usize,
+    route: String,
+}
+
+impl TimedBody {
+    fn new(inner: Body, route: String) -> Self {
+        Self {
+            inner,
+            reading_started_at: None,
+            size: 0,
+            route,
+        }
+    }
+
+    fn finish_timing(&mut self, status: &str) {
+        if let Some(started_at) = self.reading_started_at.take() {
+            self.emit_metric(started_at.elapsed(), status);
+        }
+    }
+
+    fn emit_metric(&self, duration: Duration, status: &str) {
+        metric!(
+            timer(RelayTimers::BodyReading) = duration,
+            route = &self.route,
```

Comment:
We had to turn this tag off as it was creating an extremely high cardinality metric: 1860613 time series.